perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -974,3 +976,158 @@ def MoveFile(self, target, source_path, remote_path=''):
 
     # Copies the file to its final destination in the container
     target.ContainerCopy(file_name, remote_path)
+
+
+class JujuMixin(DebianMixin):
+  """
+  Class to allow running juju-deployed workloads
+
+  Bootstraps a Juju environment using the manual provider:
+  https://jujucharms.com/docs/stable/config-manual
+  """
+
+  OS_TYPE = os_types.JUJU
+
+  isController = False
+
+  """
+  A reference to the juju controller, useful when operations occur against
+  a unit's VM but need to be preformed from the controller.
+  """
+  controller = None
+
+  units = []
+
+  environments_yaml = """
+  default: perfkit
+
+  environments:
+      perfkit:
+          type: manual
+          bootstrap-host: {0}
+  """
+
+  def _Bootstrap(self):
+    resp, _ = self.RemoteHostCommand("juju bootstrap")
+
+  def JujuAddMachine(self, ip):
+    resp, _ = self.RemoteHostCommand("juju add-machine ssh:%s" % ip)
+
+  def JujuConfigureEnvironment(self):
+    if self.isController:
+      resp, _ = self.RemoteHostCommand("mkdir -p ~/.juju")
+      f = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
FYI we have a vm_util.NamedTemporaryFile that can be used with a "with" statement and still be closed inside the statement. I'd recommend using that